modules/core/src/main/java/org/apache/ignite/internal/util/typedef/X.java

```diff
@@ -464,37 +469,26 @@ private static void cloneField(Map<Integer, Integer> identityIdxs, List<Object>
      * @return {@code True} if one of the causing exception is an instance of passed in classes,
      *      {@code false} otherwise.
      */
-    @SafeVarargs
-    public static boolean hasCause(@Nullable Throwable t, @Nullable String msg, @Nullable Class<?>... cls) {
+    public static boolean hasCause(
+        @Nullable Throwable t,
+        @Nullable String msg,
+        @Nullable Class<?>... cls
+    ) {
         if (t == null || F.isEmpty(cls))
             return false;
 
-        assert cls != null;
-
-        for (Throwable th = t; th != null; th = th.getCause()) {
+        Predicate<Throwable> pred = thr -> {
```

Comment:
Use method instead predicate, for use in hot methods.

Reply:
Are you trying to reduce `dynamic dispatch` byte-code instruction in this case?
We have to use a different `traverse` function in each case(functions cause() and hasCause() demand different predicates) - that yields copy-paste in algorithm.
I am not sure that a good thing without benchmarking because the creation throwable itself is a very expensive process (mostly capturing stack trace) and `dynamic dispatch` overhead is not so visible against it.

Reply:
Alternatively, we can traverse the whole 'tree' of throwables and memorize all elements in some list, but we will lose short-circuit`ness of operation and more garbage will be produced.

Reply:
I've fixed this possible issue by inlining the filter into the traversal function. I don't think we need to expand filters in future - so it looks ok in possibly hot code.